src/Components/Components/src/Routing/Router.cs

```diff
@@ -194,51 +195,85 @@ internal virtual void Refresh(bool isNavigationIntercepted)
                 return true;
             }
 
-            // If we've already invoked a task and stored its CTS, then
-            // cancel that existing CTS.
+            // Cancel the CTS instead of disposing it, since disposing does not
+            // actually cancel and can cause unintended Object Disposed Exceptions.
+            // This effectivelly cancels the previously running task and completes it.
             _onNavigateCts?.Cancel();
-            // Then make sure that the task has been completed cancelled or
-            // completed before continuing with the execution of this current task.
+            // Then make sure that the task has been completely cancelled or completed
+            // before starting the next one. This avoid race conditions where the cancellation
+            // for the previous task was set but not fully completed by the time we get to this
+            // invocation.
             await previousOnNavigate;
 
-            // Create a new cancellation token source for this instance
             _onNavigateCts = new CancellationTokenSource();
             var navigateContext = new NavigationContext(path, _onNavigateCts.Token);
 
-            // Create a cancellation task based on the cancellation token
-            // associated with the current running task.
             var cancellationTcs = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
             navigateContext.CancellationToken.Register(state =>
                 ((TaskCompletionSource)state).SetResult(), cancellationTcs);
 
             var task = OnNavigateAsync.InvokeAsync(navigateContext);
 
-            // If the user provided a Navigating render fragment, then show it.
             if (Navigating != null && task.Status != TaskStatus.RanToCompletion)
             {
                 _renderHandle.Render(Navigating);
             }
 
             var completedTask = await Task.WhenAny(task, cancellationTcs.Task);
-            return task == completedTask;
+
+            // We should not render the component associated with the route we
+            // are navigating to if:
+            // - The OnNavigateAsync task threw an exception
+            // - The OnNavigateAsync task wasn't cancelled by our CTS
+            // - The cancellation task completes first
+            if (task == completedTask)
+            {
+                if (completedTask.IsCompletedSuccessfully)
+                {
+                    return true;
+                }
+
+                try
+                {
+                    await completedTask;
+                }
+                catch (Exception ex)
+                {
+                    if (completedTask.IsCanceled && !_onNavigateCts.IsCancellationRequested)
+                    {
+                        throw new InvalidOperationException("OnNavigateAsync callback cannot be canceled.", ex);
+                    }
+                    ThrowExceptionInRenderer(completedTask);
```

Comment:
Good catch! This is cruft from a previous model where we checked to see if the task raised an exception using `if (task.IsFaulted)` instead of a `try/catch`.